lib/os/p4wq.c

```diff
@@ -88,7 +88,17 @@ static FUNC_NORETURN void p4wq_loop(void *p0, void *p1, void *p2)
 			thread_clear_requeued(_current);
 
 			k_spin_unlock(&queue->lock, k);
+
+			/*
+			 * No race here - we're protected by w->thread != NULL,
+			 * but the handler can resubmit and reset w->thread to
+			 * NULL, so we have to hold the mutex for the whole
+			 * duration of the handler.
+			 */
+
+			k_mutex_lock(&w->done_mutex, K_FOREVER);
```

Comment:
Really not liking this.  You can't take a blocking primitive in the main loop, it will stall the whole queue.  Consider a semaphore if you need to have user threads waiting, that can be given without blocking.

Reply:
in principle I understand your argument, but have a look what it's racing against. `k_p4wq_wait()` below really cannot hold the mutex for any significant time. And it's a per-work item mutex. So, I don't think this can create any real congestion. As for the semaphore - sure, in principle I could do that, but a semaphore is a counting primitive, which isn't what we need here.
Besides, look at `k_p4wq_wait()` again: if `.sync` isn't set and `k_p4wq_wait()` is called to regain ownership of the work item, without the mutex we have a race there - it can be called right after the handler resubmitted the work item and before we re-take the spin-lock. In that case `k_p4wq_wait()` can wrongly return 0, i.e. "work item belongs to us" and the caller will free it too early.

Reply:
I'll look closer to see exactly what this is trying to protect, but "no blocking primitives in a work queue main loop" is a pretty firm design rule here.  Again, I'm 99% certain that the correct IPC primitive for what you want is a semaphore (which allows another thread to block, but which doesn't ever block the thread calling k_sem_give()) and not a priority-inheriting mutex.

Reply:
@andyross ok, I think this race can also be avoided by just checking the semaphore counter, using `k_sem_count_get()` in the non-blocking ownership re-acquisition case. We would just take the semaphore in `k_p4wq_submit()` and give it in `p4wq_loop()` right at the end of the `if (r)` block. I think your `thread_was_requeued()` trick wouldn't be needed then either.